src/mantarray_desktop_app/mc_comm.py

```diff
@@ -2,17 +2,67 @@
 """Process controlling communication with Mantarray Microcontroller."""
 from __future__ import annotations
 
+import datetime
+from multiprocessing import Queue
+from time import perf_counter
 from time import sleep
 from typing import Any
 from typing import List
+from zlib import crc32
 
+import serial
+import serial.tools.list_ports as list_ports
+
+from .constants import SERIAL_COMM_ADDITIONAL_BYTES_INDEX
+from .constants import SERIAL_COMM_BAUD_RATE
+from .constants import SERIAL_COMM_CHECKSUM_FAILURE_PACKET_TYPE
+from .constants import SERIAL_COMM_CHECKSUM_LENGTH_BYTES
 from .constants import SERIAL_COMM_MAGIC_WORD_BYTES
+from .constants import SERIAL_COMM_MAIN_MODULE_ID
 from .constants import SERIAL_COMM_MAX_PACKET_LENGTH_BYTES
+from .constants import SERIAL_COMM_MODULE_ID_INDEX
+from .constants import SERIAL_COMM_PACKET_TYPE_INDEX
+from .constants import SERIAL_COMM_STATUS_BEACON_PACKET_TYPE
 from .constants import SERIAL_COMM_STATUS_BEACON_PERIOD_SECONDS
+from .exceptions import SerialCommIncorrectChecksumFromInstrumentError
+from .exceptions import SerialCommIncorrectChecksumFromPCError
+from .exceptions import SerialCommIncorrectMagicWordFromMantarrayError
 from .exceptions import SerialCommPacketRegistrationReadEmptyError
 from .exceptions import SerialCommPacketRegistrationSearchExhaustedError
 from .exceptions import SerialCommPacketRegistrationTimoutError
+from .exceptions import UnrecognizedSerialCommModuleIdError
+from .exceptions import UnrecognizedSerialCommPacketTypeError
 from .instrument_comm import InstrumentCommProcess
+from .mc_simulator import MantarrayMcSimulator
+from .serial_comm_utils import validate_checksum
+
+
+def _get_formatted_utc_now() -> str:
+    return datetime.datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S.%f")
+
+
+def _get_seconds_since_read_start(start: float) -> float:
+    return perf_counter() - start
+
+
+def _process_main_module_comm(comm_from_instrument: bytes) -> None:
+    packet_type = comm_from_instrument[SERIAL_COMM_PACKET_TYPE_INDEX]
+    if packet_type == SERIAL_COMM_CHECKSUM_FAILURE_PACKET_TYPE:
+        returned_packet = (
+            SERIAL_COMM_MAGIC_WORD_BYTES
+            + comm_from_instrument[
+                SERIAL_COMM_ADDITIONAL_BYTES_INDEX:-SERIAL_COMM_CHECKSUM_LENGTH_BYTES
+            ]
+        )
+        raise SerialCommIncorrectChecksumFromPCError(returned_packet)
+
+    if packet_type == SERIAL_COMM_STATUS_BEACON_PACKET_TYPE:
+        pass
+    else:
+        module_id = comm_from_instrument[SERIAL_COMM_MODULE_ID_INDEX]
```

Comment:
similar question as above about whether just an `IndexError` would happen.  maybe a check early on that the packet is at least the minimum expected length where things are being accessed by constant indices within the bytearray?